File: config/redis.py
from urllib.parse import quote

from .base import BaseConfig


class RedisConfig(BaseConfig):
    redis_host: str = "localhost"
    redis_port: int = 6379
    redis_name: str | int = 0
    redis_pass: str | None = None
    redis_socket_timeout: float = 5.0
    redis_socket_connect_timeout: float = 5.0

    @property
    def url(self) -> str:
        encoded_pass = quote(self.redis_pass or "")
        return f"redis://:{encoded_pass}@{self.redis_host}:{self.redis_port}/{self.redis_name}"

File: crates/config/src/redis.rs
use clap::Args;

#[derive(Args, Debug, Clone)]
pub struct RedisConfig {
    #[arg(long, env = "REDIS_HOST", default_value = "localhost")]
    pub redis_host: String,

    #[arg(long, env = "REDIS_PORT", default_value_t = 6379)]
    pub redis_port: u16,

    #[arg(long, env = "REDIS_PASS", default_value = "")]
    pub redis_pass: String,
}

impl RedisConfig {
    #[must_use]
    pub fn url(&self) -> String {
        if self.redis_pass.is_empty() {
            format!("redis://{}:{}", self.redis_host, self.redis_port)
        } else {
            format!(
                "redis://:{}@{}:{}",
                self.redis_pass, self.redis_host, self.redis_port
            )
        }
    }
}
